File: node/src/sensor/smoke.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/**
 * Property names of a smoke sensor.
 *
 * @internal
 */
export enum SmokeProperty {
  /** Whether smoke is detected. */
  Detected = 'detected',
}

/**
 * Property values of a smoke sensor.
 *
 * @internal
 */
export interface SmokeSensorProperties {
  [SmokeProperty.Detected]: boolean;
}

/** Read-only proxy interface for a smoke sensor. */
export interface SmokeSensorLike extends SensorLike {
  readonly type: SensorType.Smoke;
  readonly onPropertyChanged: Observable<PropertyChangeOf<SmokeSensorProperties>>;

  getValue(property: SmokeProperty.Detected): boolean | undefined;
  getValue(property: string): unknown;
}

/** Smoke detector sensor. */
export class SmokeSensor<TStorage extends object = Record<string, any>> extends Sensor<SmokeSensorProperties, TStorage> {
  readonly type = SensorType.Smoke;
  readonly category = SensorCategory.Sensor;

  constructor(name = 'Smoke Sensor', options?: SensorOptions) {
    super(name, options);

    this._writeState({ [SmokeProperty.Detected]: false });
  }

  get detected(): boolean {
    return this.props.detected;
  }

  /**
   * Report smoke detection state.
   *
   * @param value - True when smoke is currently detected.
   *
   * @example
   * ```ts
   * smoke.setDetected(true);
   * ```
   */
  setDetected(value: boolean): void {
    this._writeState({ [SmokeProperty.Detected]: value });
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Registry metadata for {@link SmokeSensor}. */
export const smokeMeta = defineSensor({
  type: SensorType.Smoke,
  category: SensorCategory.Sensor,
  assignmentKey: 'smoke',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [SmokeProperty.Detected]: { type: 'boolean', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: SmokeProperty.Detected, value: true, sustained: true },
  virtual: { properties: { [SmokeProperty.Detected]: false } },
  semantics: {
    domain: SensorDomain.Binary,
    stateProperty: SmokeProperty.Detected,
    commandProperty: SmokeProperty.Detected,
    deviceClass: 'smoke',
  },
});

File: python/camera_ui_sdk/sensor/smoke.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class SmokeProperty(StrEnum):
    """Property names of a smoke sensor."""

    Detected = "detected"
    """Whether smoke is detected."""


class SmokeSensorProperties(TypedDict):
    """Property values of a smoke sensor."""

    detected: bool


class SmokePropertyChangeData(TypedDict):
    """Property change payload emitted on SmokeSensorLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a SmokeProperty value."""
    value: bool
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class SmokeSensorLike(SensorLike, Protocol):
    """Read-only proxy interface for a smoke sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Smoke

    @property
    def onPropertyChanged(self) -> Observable[SmokePropertyChangeData]: ...

    @overload
    def getValue(self, property: Literal[SmokeProperty.Detected]) -> bool | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class SmokeSensor(Sensor[SmokeSensorProperties, TStorage, str], Generic[TStorage]):
    """Smoke detector sensor."""

    _requires_frames = False

    def __init__(self, name: str = "Smoke Sensor", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state({SmokeProperty.Detected.value: False})

    @property
    def type(self) -> SensorType:
        return SensorType.Smoke

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Sensor

    @property
    def detected(self) -> bool:
        return bool(self.props.detected)

    def setDetected(self, value: bool) -> None:
        """Report smoke detection state.

        Args:
            value: True when smoke is currently detected.

        Example:
            ```python
            smoke.setDetected(True)
            ```
        """
        self._write_state({SmokeProperty.Detected.value: value})

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored."""
